### PR TITLE
add a native C demangler

### PR DESCRIPTION
**Pull request by arielb1**
Adds a native C demangler to allow users with a build system that does not support Rust to demangle Rust symbols.

There are fuzz tests that check output compatibility between the native C demangler and the Rust demangler.

### CODE REVIEW

**Comment by arielb1** [comment]
cc @Mark-Simulacrum

**Comment by bjorn3** [comment]
How does this compare with https://github.com/LykenSol/rust-demangle.c?

**Comment by arielb1** [comment]
@Mark-Simulacrum  was not able to find the other one

**Comment by arielb1** [comment]
Personally, I think that having the demangler fuzzable against the Rust one makes me much more comfortable putting it in the rust-lang org.

**Comment by tgross35** [comment]
Fyi `crates/native-c/Cargo.toml` and `crates/native-c/README` need a trailing newline

image

File: crates/native-c/include/demangle.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+/// Beware that `res` contains references into `s`. If `s` is modified (or free'd) before calling
+/// `rust_demangle_display_demangle` behavior is undefined.
+///
+/// Use `rust_demangle_display_demangle` to convert it to an actual string.
+void rust_demangle_demangle(const char *s, struct demangle *res);
+
+/// Write the string in a `struct demangle` into a buffer.
+///
+/// Return `OverflowOk` if the output buffer was sufficiently big, `OverflowOverflow` if it wasn't.
+/// This function is `O(n)` in the length of the input + *output* [$], but the demangled output of demangling a symbol can
```
Comment (Mark-Simulacrum) [inline crates/native-c/include/demangle.h:56]: IIUC, output here refers specifically to the produced demangling, not the output buffer's size, right? We don't do anything with unused bytes in the output buffer.

(Just to check my understanding, no changes needed here I think).

File: crates/native-c/README [hunk 2]
```diff
@@ -0,0 +1,5 @@
+A portable native C demangler, which should mostly have byte-for-byte identical outputs to the Rust one.
```
Comment (Mark-Simulacrum) [inline crates/native-c/README:1]: It might be worth noting what our expectations on safety are (e.g., do we fuzz etc sufficiently that we feel confident there's no UB on untrusted inputs)? And maybe whether we also fail in the same cases as the Rust one.

File: crates/native-c/src/build.rs
Comment (Mark-Simulacrum) [inline crates/native-c/src/build.rs:1]: Accidentally added file?

File: fuzz/fuzz_targets/native_c.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            result.push(ch);
+        } else {
+            write!(&mut result, "\\u{{{:x}}}", ch as u32).ok();
+        }
+    }
+    result
+}
+
+fuzz_target!(|data: &[u8]| {
+    if data.len() == 0 {
```
Comment (Mark-Simulacrum) [inline fuzz/fuzz_targets/native_c.rs:42]: Can we add a test to make sure an empty input is OK too?

File: fuzz/fuzz_targets/native_c.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        return;
+    }
+    let alternate = data[0] % 2 == 0;
+    let data = &data[1..];
+    let mut str_buf = String::with_capacity(16384);
+    let mut buf = [0u8; 4096];
+    let mut demangle = rustc_demangle_native_c::CDemangle::zero();
+    let state;
+    if let Ok(s) = std::str::from_utf8(data) {
```
Comment (Mark-Simulacrum) [inline fuzz/fuzz_targets/native_c.rs:51]: Should we require utf-8? Or maybe we should have a separate fuzz target which just throws arbitrary input at the C impl without verifying outputs or so (or unit tests with assertions of errors on non-ascii input?)

File: fuzz/fuzz_targets/native_c.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        if let Ok(cs) = CString::new(data) {
+            unsafe {
+                rustc_demangle_native_c::rust_demangle_demangle(cs.as_ptr(), &mut demangle);
+                match rustc_demangle_native_c::rust_demangle_display_demangle(
+                    &demangle,
+                    buf.as_mut_ptr().cast(),
+                    buf.len() / 4,
```
Comment (Mark-Simulacrum) [inline fuzz/fuzz_targets/native_c.rs:58]: Why the divide by 4?

File: fuzz/fuzz_targets/native_c.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                    };
+                    let rust_overflowed = if alternate {
+                        write!(fmt_buf, "{:#}", rdemangle)
+                    } else {
+                        write!(fmt_buf, "{}", rdemangle)
+                    };
+                    if rust_overflowed.is_err() {
+                        return; // rust overflowed as well, OK
+                    }
+                    // call C again with larger buffer. If it fits in an 1020-byte Rust buffer, it will fit in a 4096-byte C buffer
```
Comment (Mark-Simulacrum) [inline fuzz/fuzz_targets/native_c.rs:90]: I'm not sure I understand why we have this special cased -- it feels like we ought to catch discrepancies of output with an assert on equivalent outputs; we don't really need differently sized buffers to catch that?

If we want to assert the overflow handling is correctly implemented maybe having a unit test that confirms output sizes less than the required output size all fail (e.g., 0..4kb for a 4kb output or so). Probably worth a test with zero-sized output buffer as well.

File: crates/native-c/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "rustc-demangle-native-c"
+version = "0.1.0"
+authors = ["automatically generated"]
+description = """
+Native C version of the rustc_demangle crate
+"""
+license = "MIT/Apache-2.0"
+repository = "https://github.com/rust-lang/rustc-demangle"
```
Comment (Mark-Simulacrum) [inline crates/native-c/Cargo.toml:9]: nit: edition = 2021

File: fuzz/Cargo.toml
Comment (Mark-Simulacrum) [inline fuzz/Cargo.toml:1]: It looks like currently CI builds but doesn't actually run fuzzing (seems like a bug in CI, to some extent) -- presumably you've run this locally? Maybe we can wire up building the C code with sanitizers and/or running fuzzing under valgrind?

File: crates/native-c/README
Comment (Mark-Simulacrum) [inline crates/native-c/README:1]: It would probably also make sense to update the top-level README noting this exists and indicating that expected usage is probably copy/paste or git submodule?

File: crates/native-c/include/demangle.h (hunk 1 above)
Comment (arielb1) [reply crates/native-c/include/demangle.h:56]: Yea, "output-dependent complexity"

File: crates/native-c/src/build.rs
Comment (arielb1) [reply crates/native-c/src/build.rs:1]: No, it's here for the fuzz harness

Comment (arielb1) [reply crates/native-c/src/build.rs:1]: Removed

File: fuzz/fuzz_targets/native_c.rs (hunk 3 above)
Comment (arielb1) [reply fuzz/fuzz_targets/native_c.rs:42]: will add

File: fuzz/fuzz_targets/native_c.rs (hunk 4 above)
Comment (arielb1) [reply fuzz/fuzz_targets/native_c.rs:51]: the C code requires ascii-only pretty early

File: fuzz/Cargo.toml
Comment (arielb1) [reply fuzz/Cargo.toml:1]: yes ran it locally

File: crates/native-c/include/demangle.h
Comment (bjorn3) [inline crates/native-c/include/demangle.h:1]: This file should probably be wrapped in

```c
#ifdef __cplusplus
extern "C" {
#endif
```

and

```c
#ifdef __cplusplus
}
#endif
```

to be able to use it from C.

File: fuzz/fuzz_targets/native_c.rs (hunk 6 above)
Comment (arielb1) [reply fuzz/fuzz_targets/native_c.rs:90]: > If we want to assert the overflow handling is correctly implemented maybe having a unit test that confirms output sizes less than the required output size all fail (e.g., 0..4kb for a 4kb output or so). Probably worth a test with zero-sized output buffer as well.

I'll add a unit test for that as well, but the fuzzing helps too

File: crates/native-c/include/demangle.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+///
+/// `alternate`, if true, uses the less verbose alternate formatting (Rust `{:#}`) is used, which does not show
+/// symbol hashes and types of constant ints.
+///
+/// [$] It's `O(n * MAX_DEPTH)`, but `MAX_DEPTH` is a constant 300 and therefore it's `O(n)`
+/// [$$] Technically, bounded by `O(n^MAX_DEPTH)`, but this is practically exponential.
+DEMANGLE_NODISCARD overflow_status rust_demangle_display_demangle(struct demangle const *res, char *out, size_t len, bool alternate);
+
+/// Returns true if `res` refers to a known valid Rust demangling style, false if it's an unknown style.
+bool rust_demangle_is_known(struct demangle *res);
```
Comment (bjorn3) [inline crates/native-c/include/demangle.h:68]: There are several functions exported by demangle.c which aren't declared in the header. For example rust_demangle_v0_demangle.

Comment (arielb1) [reply crates/native-c/include/demangle.h:68]: I'll mark them as static, it's smarter

Comment (arielb1) [reply crates/native-c/include/demangle.h:68]: They were used for testing, now they are not

File: crates/native-c/README
Comment (arielb1) [reply crates/native-c/README:1]: added it

File: crates/native-c/README (hunk 2 above)
Comment (arielb1) [reply crates/native-c/README]: added

**Review by Mark-Simulacrum — APPROVED** [review]
